Beautify readme (#442)
pygfx/pygfx: README.md

# pygfx
[![CI](https://github.com/pygfx/pygfx/workflows/CI/badge.svg)](https://github.com/pygfx/pygfx/actions)
[![Documentation Status](https://readthedocs.org/projects/pygfx/badge/?version=latest)](https://pygfx.readthedocs.io/en/latest/?badge=latest)
[![PyPI version](https://badge.fury.io/py/pygfx.svg)](https://badge.fury.io/py/pygfx)

A python render engine targeting Vulkan/Metal/DX12.

<p align="center">
<img src="./docs/_static/readme_sponza.png" alt="drawing" width="200"/>
<img src="./docs/_static/readme_pbr_example.webp" alt="drawing" width="200"/>
<img src="./docs/_static/readme_torus_knot_wire.png" alt="drawing" width="200"/><br />
</p>
<p align="center">
(Check out the <a href="https://pygfx.readthedocs.io/en/latest/_gallery/index.html">example gallery</a> for more renders.)
</p>

## Installation

If you are just starting out, or have no requirement on the canvas use

```bash
pip install -U pygfx glfw
```

If you already have a wgpu-compatible canvas (e.g., Qt, pyside, wx) installed
you may choose to not install glfw. Similarly, if you only wish to use offscreen

> rendering (eg. in CI) you will also not need glfw.

Reviewer: Maybe also mention jupyter notebook setup?

> If you already have a wgpu-compatible canvas (e.g., Qt, pyside, wx) installed
> you may choose to not install glfw. Similarly, if you only wish to use offscreen
> rendering (eg. in CI) you will also not need glfw.

Reviewer 2: Since glfw is really light there is not a problem if its installed "extra", so I'd put less emphasis on the not-installing part. Something along the lines of:

```suggestion
The glfw library provides a lightweight window to render to. It's not strictly needed
if you only use pygfx from the notebook, use offscreen rendering, or plan to use 
another wgpu-compatible canvas such as PyQt, PySide, or wx.
```

[![CI ](https://github.com/pygfx/pygfx/workflows/CI/badge.svg)
](https://github.com/pygfx/pygfx/actions)
[![Documentation Status
](https://readthedocs.org/projects/pygfx/badge/?version=latest)
](https://pygfx.readthedocs.io/en/latest/?badge=latest)
[![PyPI version ](https://badge.fury.io/py/pygfx.svg)
](https://badge.fury.io/py/pygfx)

A python render engine targeting Vulkan/Metal/DX12.

<p align="center">
<img src="./docs/_static/readme_sponza.png" alt="drawing" width="200"/>
<img src="./docs/_static/readme_pbr_example.webp" alt="drawing" width="200"/>
<img src="./docs/_static/readme_torus_knot_wire.png" alt="drawing" width="200"/>
</p>
<p align="center">
[<a href="https://pygfx.readthedocs.io/en/latest/guide.html">User Guide</a>]
[<a href="https://pygfx.readthedocs.io/en/latest/_gallery/index.html">Example Gallery</a>]
[<a href="https://pygfx.readthedocs.io/en/latest/reference.html">API Reference</a>]
</p>

## Installation

```bash
pip install -U pygfx glfw
```

To work correctly, pygfx needs _some_ window to render to. Glfw is one
lightweight option, but there are others, too. If you only use pygfx for

> offscreen, or notebook rendering you may choose to omit glfw. Similarly, you can

Reviewer 2: ```suggestion
offscreen or notebook rendering you may choose to omit glfw. Similarly, you can
```

(of course) swap it for any other wgpu-compatible canvas, e.g., PyQt, PySide, or
wx.

In addition there are some [platform
requirements](https://github.com/pygfx/wgpu-py#platform-requirements). In
essense, you need modern (enough) graphics drivers, and `pip >= 20.3`.

## Usage Example

> **Note**
> A walkthrough of this example can be found in [the
> guide](https://pygfx.readthedocs.io/en/latest/guide.html#how-to-use-pygfx).

```python
import pygfx as gfx

cube = gfx.Mesh(
    gfx.box_geometry(200, 200, 200),
    gfx.MeshPhongMaterial(color="#336699"),
)

def animate():
    rot = gfx.linalg.Quaternion().set_from_euler(
            gfx.linalg.Euler(0, 0.01)
        )
    cube.rotation.multiply(rot)

if __name__ == "__main__":
    gfx.show(cube, before_render=animate)

```
<img src="./docs/_static/guide_rotating_cube.gif" alt="drawing" width="400"/>

## Feature Highlights
Some of pygfx's key features are:

> - SDF based rendering (example)

Reviewer 2: ```suggestion
- SDF based text rendering (example)
```

- order-independent transparency (OIT) (example)

> - a shader templating system (example)

Reviewer 2: The shader templating is cool, but only interesting for ppl that want to extend pygfx. Let's replace with lights etc.
```suggestion
- lights, shadows, and physically based rendering (PBR)
```

- event system with built-in picking (example)
- Texture and color mapping supporting 1D, 2D and **3D data** (example)

And many more! Check out our [feature
demos](https://pygfx.readthedocs.io/en/latest/_gallery/index.html) in the docs.

## About pygfx

> Pygfx is a graphics library that uses WGPU (the successor of OpenGL) to provide

Author: @ivoflipse This one is at the beginning of a new sentence (paragraph in fact) so I capitalized it. It feels strange not to capitalize, though we can ofc use lowercase if you prefer.

> Pygfx is a graphics library that uses WGPU (the successor of OpenGL) to provide

Reviewer 3: What does scikit-learn do for example?

> Pygfx is a graphics library that uses WGPU (the successor of OpenGL) to provide

Reviewer 3: It seems to vary, which probably depends on who wrote it

https://scikit-learn.org/stable/faq.html

> Pygfx is a graphics library that uses WGPU (the successor of OpenGL) to provide

Author: On second thought, let's avoid starting a long discussion over something minor. I'll change it.

> GPU acceleration to rendering workloads. It mimics the ThreeJS API and is geared

Reviewer 2: It's more "inspired by" than "mimics", because we (deliberately) in some cases :)

towards scientific and medical visualization. At the same time, is mature enough

> to serve as a general-purpose rendering engine. (Yes, you _can_ write a game

Reviewer 2: Can we flip this around - first mention it's general purpose, then say it's geared towards scientific vis?

Author: swapped

> with it.) Thanks to it's low level of abstraction it is flexible and can be

Author: fixed

adapted to various use-cases. In other words, pygfx emphasizes on hackability
and correctness while maintaining the level of performance you would expect from
a GPU accelerated library.

> @almarklein: ? citation ?

Reviewer 2: What kind of citation do you mean?

Reviewer 2: Let's skip for now.

## License

Pygfx is licensed under the [BSD 2-Clause "Simplified" License](LICENSE). This means:

- :white_check_mark: It is free (and open source) forever. :cupid:
- :white_check_mark: You _can_ use it commercially.
- :white_check_mark: You _can_ distribute it and freely make changes.
- :x: You _can not_ hold us accountable the results of using pygfx.

## Contributing
We use a pull request (PR) based workflow similar to many other open-source
libraries in the python ecosystem. You can read more about this workflow

> [here](https://docs.github.com/en/get-started/quickstart/github-flow); however,

Reviewer 2: The next sentence is more a consequence than a contradiction
```suggestion
[here](https://docs.github.com/en/get-started/quickstart/github-flow);
```

[![Documentation Status
](https://readthedocs.org/projects/pygfx/badge/?version=latest)
](https://pygfx.readthedocs.io/en/latest/?badge=latest)
[![PyPI version ](https://badge.fury.io/py/pygfx.svg)
](https://badge.fury.io/py/pygfx)

A python render engine targeting Vulkan/Metal/DX12.

<p align="center">
<img src="./docs/_static/readme_sponza.png" alt="drawing" width="200"/>
<img src="./docs/_static/readme_pbr_example.webp" alt="drawing" width="200"/>
<img src="./docs/_static/readme_torus_knot_wire.png" alt="drawing" width="200"/>
</p>
<p align="center">
[<a href="https://pygfx.readthedocs.io/en/latest/guide.html">User Guide</a>]
[<a href="https://pygfx.readthedocs.io/en/latest/_gallery/index.html">Example Gallery</a>]
[<a href="https://pygfx.readthedocs.io/en/latest/reference.html">API Reference</a>]
</p>

## Installation

```bash
pip install -U pygfx glfw
```

To work correctly, pygfx needs _some_ window to render to. Glfw is one
lightweight option, but there are others, too. If you use a different
wgpu-compatible window manager or only render offscreen you may choose to omit
glfw. Examples of alternatives include: jupyter_rfb (rendering in jupyter),
PyQt, PySide, or wx.

> In addition there are some [platform
> requirements](https://github.com/pygfx/wgpu-py#platform-requirements). In
> essense, you need modern (enough) graphics drivers, and `pip>=20.3`.

Reviewer 2: These platform requirements in wgpu-py have moved into its docs.

```suggestion
In addition there are some platform
requirements, see the [wgpu docs](https://wgpu-py.readthedocs.io/en/stable/start.html). In
essence, you need modern (enough) graphics drivers, and `pip>=20.3`.
```

## Usage Example

> **Note**
> A walkthrough of this example can be found in [the
> guide](https://pygfx.readthedocs.io/en/latest/guide.html#how-to-use-pygfx).

```python
import pygfx as gfx

cube = gfx.Mesh(
    gfx.box_geometry(200, 200, 200),
    gfx.MeshPhongMaterial(color="#336699"),
)

def animate():
    rot = gfx.linalg.Quaternion().set_from_euler(
            gfx.linalg.Euler(0, 0.01)
        )
    cube.rotation.multiply(rot)

if __name__ == "__main__":
    gfx.show(cube, before_render=animate)

```
<img src="./docs/_static/guide_rotating_cube.gif" alt="drawing" width="400"/>

## Feature Highlights
Some of pygfx's key features are:

- SDF based text rendering ([example](
  https://pygfx.readthedocs.io/en/latest/_gallery/feature_demo/text_contrast.html))
- order-independent transparency (OIT) ([example](
  https://pygfx.readthedocs.io/en/latest/_gallery/feature_demo/transparency2.html))
- lights, shadows, and physically based rendering (PBR) ([example](
  https://pygfx.readthedocs.io/en/latest/_gallery/feature_demo/pbr.html))
- event system with built-in picking ([example](
  https://pygfx.readthedocs.io/en/latest/_gallery/feature_demo/picking_points.html))

> - Texture and color mapping supporting 1D, 2D and **3D data** ([example](

Reviewer 2: Consistency.
```suggestion
- texture and color mapping supporting 1D, 2D and 3D data ([example](
```

Author: nice catch.

# pygfx
[![CI ](https://github.com/pygfx/pygfx/workflows/CI/badge.svg)
](https://github.com/pygfx/pygfx/actions)
[![Documentation Status
](https://readthedocs.org/projects/pygfx/badge/?version=latest)
](https://pygfx.readthedocs.io/en/latest/?badge=latest)
[![PyPI version ](https://badge.fury.io/py/pygfx.svg)
](https://badge.fury.io/py/pygfx)

A python render engine targeting Vulkan/Metal/DX12.

<p align="center">
<img src="./docs/_static/readme_sponza.png" alt="drawing" width="200"/>
<img src="./docs/_static/readme_pbr_example.webp" alt="drawing" width="200"/>
<img src="./docs/_static/readme_torus_knot_wire.png" alt="drawing" width="200"/>
</p>
<p align="center">
(Check out the 
<a href="https://pygfx.readthedocs.io/en/latest/_gallery/index.html"> example 
gallery </a> for more renders.)
</p>

## Installation

```bash
pip install -U pygfx glfw

> ```

Reviewer: all the same caveats apply as for wgpu-py, e.g. you need pip >=20.3 on linux, etc. I think that's going to lead to the same kind of issues on this repo if we don't mention it like we do over there

Author: I added a link to the platform requirements for wgpu-py and a general note.

Pygfx needs _some_ window to render to. Glfw is one lightweight option, but
others do exist. If you only use pygfx for offscreen, or notebook rendering you
may choose to omit glfw. Similarly, you can (of course) swap it for any other
wgpu-compatible canvas, e.g., PyQt, PySide, or wx.

## Example

> **Note**
> A walkthrough of this example can be found in [the
> guide](https://pygfx.readthedocs.io/en/latest/guide.html#how-to-use-pygfx).

```python
import pygfx as gfx

cube = gfx.Mesh(
    gfx.box_geometry(200, 200, 200),
    gfx.MeshPhongMaterial(color="#336699"),
)

def animate():
    rot = gfx.linalg.Quaternion().set_from_euler(
            gfx.linalg.Euler(0, 0.01)
        )
    cube.rotation.multiply(rot)

if __name__ == "__main__":
    gfx.show(cube, before_render=animate)

```
<img src="./docs/_static/guide_rotating_cube.gif" alt="drawing" width="400"/>

## Features
Some of pygfx's key features are:

- Native support for High Resolution screens.
- Builtin anti-aliasing.

> - Support for picking objects and parts within objects.

Reviewer: Let's make that "Builtin picking"

@almarklein: A few suggestions and feature discussions
- fast draws (57 FPS @ 30M vertices from 1k lines on a . . .) (@almarklein: what
  specs does your laptop have? Also, do we have a polygon count benchmark?)
- (approximate) order-independent transparency (OIT). (is this implemented now?)
- Custom post-processing. (do the examples for this need updating?)
- Easy shader templating
- built atop [WGPU](https://github.com/pygfx/wgpu-py) (>= OpenGL)

> - semantically similar to TreeJS (WorldObjects, decoupled rendering backend, ...)

Reviewer: I think the event system is also worth a mention

Reviewer 2: Idea: perhaps we can split out features particular of interest for 1) object rendering / meshes; 2) scientific viz.

Author: I would only list the 3 or 5 features that we are most proud of (or that sell the best) otherwise it quickly becomes a boring cloth-line.

For all the others, we should already have a feature example that shows off each feature right? So we can just point users to the gallery and say "check if the feature you want is listed here".

Author: I've made a selection of the top 5 features from all our suggestions. I kicked out some (light/shaddows, PBR, ...) that you could already see in the example images at the top, or that is mentioned elsewhere (TreeJS/WGPU) and then choose the 5 items that sounded the most impressive.

Suggestions are, ofc, welcome, and we also have to find the corresponding examples for each feature.

## License

Pygfx is licensed under the [BSD 2-Clause "Simplified" License](LICENSE). This means:

- :white_check_mark: It is free (and open source) forever. :cupid:
- :white_check_mark: You _can_ use it commercially.
- :white_check_mark: You _can_ distribute it and freely make changes.
- :x: You _can not_ hold us accountable the results of using pygfx.

## Contributing
We use a very similar setup to many other open-source libraries in the python
ecosystem. If you have contributed to open-source in the past, a lot of this
should look familiar.

### GitHub Workflow
We use a PR-based workflow. Each contributor has his/her own fork of the project
inside of which he/she maintains a branch containing the changes that should be
incorporated into the project. These changes are presented as a [pull request
(PR)](https://github.com/pygfx/pygfx/pulls) on GitHub where we discuss them,
review code, and eventually merge the work. 

When you begin working on pygfx you would typically do something like:

```bash
# Click the Fork button on GitHub and navigate to your fork
git clone <address_of_your_fork>
cd pygfx
git remote add upstream git@github.com:pygfx/pygfx.git

> pip install -e .[dev,docs,examples]

Reviewer: Are contributors smart enough to use a venv themselves?

Author: I agree there are many existing ones, but I don't know a good one (never searched). If you know one I'd be happy to link it. It's mainly for people that are new to open-source work to have something easy to get them started.

Alternatively, we could drop the contributing section entirely or shrink it. Depends on how eager we are to attract external contributions.

Reviewer 2: This one perhaps? https://docs.github.com/en/get-started/quickstart/github-flow

Author: Sounds good. I will still keep a snippet on how to get a dev install, because that is specific to pygfx (i.e. which optional dependencies to install).
